### PR TITLE
Boost: Add Cloud CSS showstopper error

### PR DESCRIPTION
**Pull request by karthikax**
Fixes 38-gh-Automattic/boost-shield-api

#### Changes proposed in this Pull Request:

* Cloud CSS status store is refactored to align with Critical CSS status store
* Display errors on the following conditions:
  * Cloud CSS generation request to WordPress.com API is timed out.
  * Cloud CSS is not returned after a defined time (Currently set to one hour).
  * Cloud CSS generation fails with a showstopper error.

#### Jetpack product discussion

NA

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Define constant `JETPACK_BOOST_CLOUD_CSS` as true to enable Cloud Critical CSS.
* Navigate to Jetpack boost dashboard: `wp-admin/admin.php?page=jetpack-boost`
* Enable `Automatically Optimize CSS Loading`
* Confirm that Critical CSS is generated.
* Enable Jetpack Sandbox with patch `D78168`
* Simulate a request timeout or error to WordPress.com API sandbox
* Regenerate Cloud CSS by clicking the `regenerate` button
* Confirm that the appropriate error is displayed under the `Automatically Optimize CSS Loading` meta section.
* Simulate Cloud CSS never returning via `WPCOM_REST_API_V2_Endpoint_Jetpack_Boost_Cloud_CSS::update_cloud_css`
* Regenerate Cloud CSS by clicking the `regenerate` button
* Confirm that the timeout error is displayed under the `Automatically Optimize CSS Loading` meta section.
* Simulate a showstopper error
* Regenerate Cloud CSS by clicking the `regenerate` button
* Confirm that the appropriate error is displayed under the `Automatically Optimize CSS Loading` meta section.
* Disable Cloud CSS by setting `JETPACK_BOOST_CLOUD_CSS` as `false`.
* Ensure that local Critical CSS works as expected

Note: This PR needs to be tested in conjunction with D78168-code and latest Hydra

### CODE REVIEW

File: projects/plugins/boost/app/assets/src/js/stores/critical-css-status.ts [hunk 1]
```diff
@@ -169,6 +188,23 @@ export function storeGenerateError( error: Error ): void {
 	} ) );
 }
 
+export function resetCloudStatus(): void {
```
Comment (eAdnan007) [inline projects/plugins/boost/app/assets/src/js/stores/critical-css-status.ts:191]: Should this too be a generic name that would represent both? The only different thing here I notice is `status: 'requesting'`. Which shouldn't be a problem.

**Review by eAdnan007 — CHANGES_REQUESTED** [review]
I have added a few points. Kudos for merging the `Generators`. Let me know if any of my suggestion/comment does not make sense. This is a big change and I may have gotten things wrong.

File: projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php [hunk 2]
```diff
@@ -107,37 +107,11 @@ private function get_state_transient() {
...
-		$pending   = false;
-
-		foreach ( $providers as $provider ) {
-			if ( self::SUCCESS === $provider['status'] ) {
-				$completed++;
-			} elseif ( self::REQUESTING === $provider['status'] ) {
-				$pending = true;
-			}
+	public function maybe_set_status() {
+		if ( $this->get_total_providers_count() === $this->get_processed_providers_count() ) {
```
[the remark this answers is not in the document]
Comment (karthikax) [reply projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php:111]: This actually looks like a correct way to do it. However, the current implementation shows a primary error set in case of all the providers fail with non-showstopper errors and the overall status is `success`. I think we can refactor this with the scope of a new ticket.

File: projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php [hunk 3]
```diff
@@ -28,16 +28,29 @@ public function __construct( $state ) {
...
 		$this->reset_existing_css();
 		return $client->post( 'cloud-css', $payload );
 	}
 
+	private function get_css_generate_providers( $providers ) {
+		$formatted = array();
+		foreach ( $providers as $key => $urls ) {
+			$generate_urls = array();
+			foreach ( $urls as $url ) {
+				$generate_urls[] = add_query_arg( 'jb-generate-critical-css', 'true', $url );
```
Comment (eAdnan007) [inline projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php:47]: I have just noticed something. Because the local and cloud critical CSS handle adding the parameter in a different way, cloud CSS doesn't remove the parameter while displaying the error. I think we should remove it before displaying. But, it is of lower priority and we can do it in a different PR.

Btw, what's the advantage of having this param for CCSS?

**Review by eAdnan007 — APPROVED** [review]
Just a final note to consider but good to merge 👍🏼

File: projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php (hunk 3 above)
Comment (karthikax) [reply projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php:47]: You are right. This is only used for testing purposes. I do not see the necessity to keep this or at least we need to remove the arg. However, didn't want to keep this open for long, hence merged. I will update it with the advanced recommendation PR.

Comment (karthikax) [reply projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php:47]: This is now included in #23919
